dnserver/main.py

```diff
@@ -87,10 +87,12 @@ def __str__(self):
 
 class Resolver(ProxyResolver):
     def __init__(self, zones_file: str | Path, upstream: str):
+        self.upstream = upstream
         records = load_records(zones_file)
         self.records = [Record(zone) for zone in records.zones]
         logger.info('loaded %d zone record from %s', len(self.records), zones_file)
-        super().__init__(address=upstream, port=53, timeout=5)
+        if upstream:
+            super().__init__(address=upstream, port=DEFAULT_PORT, timeout=5)
```

Comment:
You have to always call `super().__init__`, otherwise nasty things can happen.

If `super().__init__` creates a socket/connection, you'll need two `Resolver` classes, with a mixin for shared logic.

If not, you can just always call `super().__init__` with the default `upstream`.